### PR TITLE
Make Swift 5.2 the minimum requirement

### PR DESCRIPTION
**Pull request by Davidde94**
Make Swift 5.2 the minimum requirement, dropping support for Swift 5.0 and 5.1.

### Motivation:

Whenever we have problems, Swift 5.0 and 5.1 seem to be the culprits. Dropping support for these very old versions will require less maintenance and free up our time to work on new features.

### Modifications:

- Set the tools version in Package.swift to 5.2
- Remove CI configurations for 5.0 and 5.1
- Update the various readmes to reflect that this change will be rolled out in NIO 2.30.0

### Result:

Swift 5.2 is the minimum version of Swift required to use NIO.

### CODE REVIEW

File: docs/migration-guide-NIO1-to-NIO2.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ This migration guide is our recommendation to migrate from NIO 1 to NIO 2. None
 
 This repository also contains a fairly complete list of public [API changes](https://github.com/apple/swift-nio/blob/main/docs/public-api-changes-NIO1-to-NIO2.md) from NIO 1 to NIO 2.
 
+### Note: As of NIO 2.30.0 the minimum version of Swift required is 5.3.
```
Comment (0xTim) [inline docs/migration-guide-NIO1-to-NIO2.md:7]: Typo?

```suggestion
### Note: As of NIO 2.30.0 the minimum version of Swift required is 5.2.
```

Comment (Davidde94) [reply docs/migration-guide-NIO1-to-NIO2.md:7]: Oops, thanks!

File: dev/make-single-file-spm [hunk 2]
```diff
@@ -51,7 +51,7 @@ function setup_swiftpm_package() {
 
     cd "$destination"
     cat > Package.swift <<"EOF"
-// swift-tools-version:5.1
+// swift-tools-version:5.2
```
Comment (Lukasa) [inline dev/make-single-file-spm:54]: Did you validate this script works? I suspect it now produces a broken SPM Package manifest.

**Review by Lukasa — CHANGES_REQUESTED** [review]
This will need to be coordinated with @tomerd to turn off the relevant builders for SwiftNIO.

File: dev/make-single-file-spm (hunk 2 above)
Comment (weissi) [reply dev/make-single-file-spm:54]: we also don't need this change. You can still depend on packages that have newer requirements with older manifests.

Comment (Davidde94) [reply dev/make-single-file-spm:54]: Good point

**Comment by Davidde94** [comment]
@Lukasa yep! He's in the review list and I'm sending an email

**Review by glbrntt — DISMISSED** [review]
I think a few other things need updating; `scripts/build_podspecs.sh` and `CONTRIBUTING.md`.

Also I think we should be able to get rid of the generated tests now and use `--enable-test-discovery` when it's necessary. Might be easier to do that in a follow up.

**Comment by weissi** [comment]
> Also I think we should be able to get rid of the generated tests now and use `--enable-test-discovery` when it's necessary. Might be easier to do that in a follow up.

Sadly not: https://bugs.swift.org/browse/SR-12008

**Comment by Davidde94** [comment]
@glbrntt we can't get rid of linux tests right now

**Comment by glbrntt** [comment]
> > Also I think we should be able to get rid of the generated tests now and use `--enable-test-discovery` when it's necessary. Might be easier to do that in a follow up.
> 
> Sadly not: https://bugs.swift.org/browse/SR-12008

euuuughhhhh 😡😡😡

**Review by Lukasa — APPROVED** [review]

**Comment by adam-fowler** [comment]
Is the setting of the Package minimum requirements something that should be coordinated with some of the key serverside swift package owners. I doubt any of them want to restrict what version of SwiftNIO their clients can use. Some obvious packages would be AHC, Soto, MongoSwift and RediStack. I think all the Vapor packages have a minimum requirement of 5.2 already.

**Comment by adam-fowler** [comment]
> Is the setting of the Package minimum requirements something that should be coordinated with some of the key serverside swift package owners. I doubt any of them want to restrict what version of SwiftNIO their clients can use. Some obvious packages would be AHC, Soto, MongoSwift and RediStack. I think all the Vapor packages have a minimum requirement of 5.2 already.

After talking to @fabianfett  offline it appears this isn't an issue. If someone is compiling with Swift 5.2 or later they will pick up later versions of SwiftNIO even though a library in the middle says its minimum requirement is Swift 5.1. Unless that library indicates it wants version 2.30 or later of NIO.

**Comment by finestructure** [comment]
In case you haven't seen the Twitter thread, here's a list of packages in the Swift Package Index that depend on NIO showing their Swift version compatibility as we compute it: https://www.icloud.com/numbers/049dkCvCOid9P8S0UxtCGE3iA#NIO-dependency-check

([Twitter thread](https://twitter.com/_sa_s/status/1398286782807035910?s=21))

Note that of the 101 packages (of 298) that are _not_ 5.2+ compatible quite a few are probably only flagged as incompatible because they use OpenSSL and we don't have it installed on our builders. I.e. there's a good chance the impact is smaller than that.

**Review by fabianfett — APPROVED** [review]
🚀🎉

File: README.md [hunk 3]
```diff
@@ -20,6 +20,8 @@ Repository | NIO 2 (Swift 5+) | NIO 1 (Swift 4+)
 [https://github.com/apple/swift-nio-transport-services][repo-nio-transport-services] <br> first-class support for macOS, iOS, tvOS, and watchOS | `from: "1.0.0"` | `from: "0.1.0"`
 [https://github.com/apple/swift-nio-ssh][repo-nio-ssh] <br> SSH support | `.upToNextMinor(from: "0.2.0")` | _n/a_
 
+NIO 2.29.0 and older support Swift 5.0+.
```
Comment (PeterAdams-A) [inline README.md:23]: Wondering if there is a better way of representing this.  5.0 would work if they did `from: 2.0.0`

**Review by PeterAdams-A — APPROVED** [review]
OK, one comment but I don't have a concrete suggestion that's better.

File: README.md (hunk 3 above)
Comment (Lukasa) [reply README.md:23]: 5.0 would work if they did `from`: anything 2.29.0 and earlier. I think this is probably sufficient.

**Comment by Davidde94** [comment]
@swift-server-bot test this please

**Comment by yim-lee** [comment]
@swift-nio-bot test this please

**Comment by yim-lee** [comment]
@swift-nio-bot test this please

**Comment by helje5** [comment]
> Whenever we have problems, Swift 5.0 and 5.1 seem to be the culprits. 
In what ways?

**Comment by helje5** [comment]
A few issues raised in the Slack:

1. https://github.com/apple/swift-nio/pull/1647 - just affects building Windows w/ <5.2, no other platforms affected. Windows affected neither, just needs to use 5.2 there
2. https://github.com/apple/swift-nio/pull/1820 - perf only, could use different compile paths for the two
3. https://github.com/apple/swift-nio/pull/1823 - just a test setup affected, can be worked around

**Pull request merged**